tests/strictness-checks/Partial.hs

```diff
@@ -0,0 +1,170 @@
+module Partial where
+
+import Control.DeepSeq
+import Control.Exception
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.Reader
+import Control.Monad.Trans.State.Strict
+import Data.Bifunctor
+import Data.Bitraversable
+import qualified Data.List as List
+import Numeric.Natural
+import Test.QuickCheck (Arbitrary (arbitrary, shrink), Gen, frequency)
+
+import Prelude'
+
+-- | A specification for a value that might not be defined.
+newtype Specification value = Specification {partial :: Maybe value} deriving (Show, Functor, Foldable, Traversable)
+pattern Undefined :: Specification value
+pattern Undefined = Specification Nothing
+pattern Defined :: value -> Specification value
+pattern Defined value = Specification (Just value)
+
+instance {-# overlappable #-} Arbitrary value => Arbitrary (Specification value) where
+  arbitrary = generatePartial arbitrary
+  shrink (Specification (Just value)) = fmap (Specification . Just) (shrink value)
+  shrink (Specification Nothing) = [ ]
+
+generatePartial :: Gen value -> Gen (Specification value)
+generatePartial generatorOfValues = frequency
+      [ (1, return (Specification Nothing))
+      , (9, fmap (Specification . Just) generatorOfValues) ]
+
+-- | Extract an error from a value that might be somewhere undefined.
+save :: NFData value => value -> IO (Either String value)
+save = fmap (first errorMessage) . try @ErrorCall . evaluate . force
+
+-- | Total values are returned as they are. They are never looked into.
+newtype Total value = Total {total :: value} deriving (Functor, Foldable, Traversable, Arbitrary)
+instance Show value => Show (Total value)
+  where
+    show = show . total
+
+type family MapPartial (typeConstructor :: * -> *) value where
+  MapPartial _ (Total value) = Total value
+  MapPartial typeConstructor (Specification value) = typeConstructor (MapPartial typeConstructor value)
+  MapPartial typeConstructor (trifunctor first second third) = trifunctor (MapPartial typeConstructor first) (MapPartial typeConstructor second) (MapPartial typeConstructor third)
+  MapPartial typeConstructor (bifunctor first second) = bifunctor (MapPartial typeConstructor first) (MapPartial typeConstructor second)
+  MapPartial typeConstructor (functor value) = functor (MapPartial typeConstructor value)
+  MapPartial _ anythingElse = anythingElse
```

Comment:
Could it be possibly better to return [`TypeError`](https://hackage.haskell.org/package/base-4.14.0.0/docs/GHC-TypeLits.html#t:TypeError)?

Reply:
I want ordinary types to be specifications of themselves, so there is never a type error.

I later added a `newtype Total` so that any type can specify itself, but it would be cumbersome to put it all over your huge value when you but want to specify this tiny leaf in the corner to be undefined, so there is this little convenience.

It would be more correct in principle to bring a `data family Specification` in, with separate constructors for all kinds of functors, bifunctors, profunctors… but then specifications would be so much heavier, so I chose not to do that for now.